### ISSUE TITLE
"Disable editing value" prevents also creation of value

### ISSUE THREAD

**Issue by the-other-dev** [directus/directus#23114]
### Describe the Bug

If a field in a collection is set to "Readonly"/"Disable editing value", you can't input a value on item creation because the respective input field is inactive. Marking the field additionally as required for creation makes it impossible to create the item.

![Screenshot 2024-07-26 at 13 26 15](https://github.com/user-attachments/assets/8133475a-b5cc-478f-b9c8-d5dfd7533cf8)

![Screenshot 2024-07-26 at 13 26 26](https://github.com/user-attachments/assets/c0de3954-31e9-405b-a788-8decc9810db7)

"disable editing value" sounds to me like it should prevent me from modifying _existing_ entries. Disabling also the input of new values does not seem correct.

N.B. It is possible to create a new item via the API.

### To Reproduce

* Create a collection with a field that is marked as "Readonly" / "Disable editing value"
* Try creating a new item in that collection
* The field input is inactive and no data can be entered

### Directus Version

v10.13.1

### Hosting Strategy

Self-Hosted (Docker Image)

### Database

PostgreSQL

**Comment by rijkvanzanten** [comment]
Ah that's a misleading label then yeah. Readonly is intended to make the field readonly in the app in all cases. It's not intended to be a 'only allow value to be set on creation'

**Comment by rijkvanzanten** [comment]
Any thoughts on a better label? MDN's description for `readonly` is very similar "_The Boolean readonly attribute, when present, makes the element not mutable, meaning the user can not edit the control._"

**Comment by the-other-dev** [comment]
I see. That raises multiple questions for me:

* If it is intended to be read-only in all cases, why can I set value via API? Is that a bug?
* How would I be able to set the value through the UI in case I declare the field as read-only? Is that even possible?
* How can I achieve the behavior of being able to not modify a value after creation (so read-only for editing, but when creating the item I can define a value)?

Regarding the naming for the label, I think mentioning "immutable" in the description might be a step in the right direction. Additionally, mentioning the behavior in the documentation, e.g. when discussing fields (https://docs.directus.io/app/data-model/fields.html) might help users to know what this option does.

**Comment by benhaynes** [comment]
Interesting.

* I think `Readonly` is the right term for how this works (can *not* create or edit), but agree we should update the docs, fix the API loophole, and look into adding a field note to clarify the setting.
* You should be able to use access control to set who can create or edit a field's values more granularly
* We could look into making readonly a more flexible field-level setting, that has options for specifically disabling entry on create versus edit.
* We could show a warning notice when both of these options (readonly and required) are enabled... to warn of this potential trap.

**Comment by the-other-dev** [comment]
Maybe the term "editing" in the label "Disable editing value" lead me astray and I was fixed on "editing" as in "modifying after creation". In my opinion this "edit" term is slightly ambiguous.

Either way, I would be happy with the API bugfix and clarification in the documentation.

**Comment by Xinnerpeace** [comment]
> * being able to not modify a value after creation (so read-only for editing, but when creating the item I can define a value)

This would be a practical option.
Do we currently have any way to achieve this?

**Comment by the-other-dev** [comment]
> > * being able to not modify a value after creation (so read-only for editing, but when creating the item I can define a value)
> 
> This would be a practical option. Do we currently have any way to achieve this?

If I am not mistaken is the only way through a custom extension (filter hook for `item.update` events or more specific for a certain collection) that checks if a specific field is part of the payload and reject if that is the case. You could implement this with a flow as well. Either way, this approach is only a viable solution for specific collections. Having a generic approach would make that custom extension more complex as the meta information (which field is read-only on edit only) needs to be implemented as well.

**Comment by danmaze** [comment]
If you ask me, I don't see why the field shouldn't be editable via the API. It has to be editable somehow otherwise the field would be useless and making it only editable during creation is limiting I think. As said above, access control should be sufficient for defining who can create or update the field and under what conditions. 

IMO, simply updating the label to specify that readonly here means "in app" should suffice - that way it serves more as a visual UI cue which is what it's most useful for.

**Comment by gavalierm** [comment]
+1 for "Disabled for UI editing" will be great option.

I want to have some fields open for API but not for GUI - like "slug" which i want to alter via API, or via creation, or via custom extension, but not from UI itself.

Note: Read-only values can be set on creation by its own default value?

**Comment by DevJobalia** [comment]
Finally what improvement is decided. So that I can take on the challenge.

`On creation not read only`

**Comment by actuallyforked** [comment]
> > * being able to not modify a value after creation (so read-only for editing, but when creating the item I can define a value)
> 
> This would be a practical option. Do we currently have any way to achieve this?

Yes, i actually found a little weird but solid solution, that works within the web interface and no plugin required:

1. Edit the field and make sure it is NOT readonly and maybe force to set it on creation
2. Go to 'Conditions' and add a new one
3. The condition should make the field readonly if: The ID is bigger than 0*

As the ID is not set until the item is created, the condition is false, thus allowing editing. Once the item is created, the condition is true and editing the field no longer is allowed via the web interface.

*: I have a numeric ID in this case, but if you use a string it should still work by checking if the ID is an empty string (or maybe null, havn't tried)

**Comment by gavalierm** [comment]
> > > * being able to not modify a value after creation (so read-only for editing, but when creating the item I can define a value)
> > 
> > 
> > This would be a practical option. Do we currently have any way to achieve this?
> 
> Yes, i actually found a little weird but solid solution, that works within the web interface and no plugin required:
> 
> 1. Edit the field and make sure it is NOT readonly and maybe force to set it on creation
> 2. Go to 'Conditions' and add a new one
> 3. The condition should make the field readonly if: The ID is bigger than 0*
> 
> As the ID is not set until the item is created, the condition is false, thus allowing editing. Once the item is created, the condition is true and editing the field no longer is allowed via the web interface.
> 
> *: I have a numeric ID in this case, but if you use a string it should still work by checking if the ID is an empty string (or maybe null, havn't tried)

It will be not able to edit after initial store.

If you want to edit "slug" via API or every time when title was changed you need to have this option.  But like your out of box hack :)

**Comment by Nitwel** [comment]
Let me know if I'm misunderstanding something here but this should already be possible using permissions.
By giving out create access to the field but not update access, you are ensuring that the field can only be set on creation. This also ensures this behavior is both checked on the app and api side.

In that sense, marking the field as required/read only are just app side options that have no effect on the api. The only improvement I can see here is to differentiate the read-only option between create and update.

@AlexGaillard I'll relabel this back to being a studio issue, could you provide some insight into why you relabeled this as being a engine bug?

**Comment by AlexGaillard** [comment]
> [@AlexGaillard](https://github.com/AlexGaillard) I'll relabel this back to being a studio issue, could you provide some insight into why you relabeled this as being a engine bug?

@Nitwel Not sure, I might have just read it wrong. It's all good, we can take it 👍

**Comment by Nitwel** [comment]
Got it, no worries! 😄
